File: java/DiningSavages.java
import java.util.Random;
import java.util.concurrent.Semaphore;

//Solution taken from little book of semaphores and applied to java
public class DiningSavages {

    static Semaphore mutex = new Semaphore(1);
    static Semaphore emptyPot = new Semaphore(0);
    static Semaphore fullPot = new Semaphore(0);
    static int servingSize = 10;
    static int savages = 100;
    static int servings = 10;

    public static void main(String args[]) {
        Cook cook = new Cook();
        Savage savage = new Savage();

        // init cook
        Thread cookThread = new Thread(cook);
        cookThread.setName("cook");
        // quit once all savages have eaten
        cookThread.setDaemon(true);
        cookThread.start();

        Thread[] threads = new Thread[savages];
        for (int i = 0; i < threads.length; i++) {
            threads[i] = new Thread(savage);
            threads[i].setName("Savage " + i);
            threads[i].start();
        }
    }

    static class Savage implements Runnable {
        @Override
        public void run() {
            try {
                mutex.acquire();
                // If pot is empty, notify cook
                if (servings == 0) {
                    emptyPot.release();
                    fullPot.acquire();
                    servings = servingSize;
                }
                servings -= 1;
                System.out.println("eating");
                mutex.release();

            } catch (InterruptedException e) {
                System.out.println(e.getMessage());
            }
        }
    }

    static class Cook implements Runnable {
        @Override
        public void run() {
            while (true) {
                try {
                    // wait for pot to be empty
                    emptyPot.acquire();

                    // Refill pot
                    System.out.println("Replenishing pot");

                    // Notify pot is full
                    fullPot.release();
                } catch (InterruptedException e) {
                    System.out.println(e.getMessage());
                }
            }
        }
    }
}

File: golang/dining_savages.go
package main

import (
	"fmt"
	"sync"
)

var servingSize = 10
var savages = 100
var wg sync.WaitGroup

func main() {
	potEmpty := make(chan int, 1)
	pot := make(chan int, servingSize)
	potFull := make(chan int, 1)
	for i := 0; i < servingSize; i++ {
		pot <- 1
	}
	potFull <- 1
	//Start all savages concurrently
	for i := 0; i < savages; i++ {
		wg.Add(1)
		go eat(potEmpty, pot, potFull)
	}

	//Start the cook
	cook(potEmpty, pot, potFull)
	//Make sure we wait for all savages to eat
	wg.Wait()
}

func cook(potEmpty, pot, potFull chan int) {
	for i := 0; i < (savages/servingSize)-1; i++ {
		//Wait till pot is empty
		<-potEmpty
		//Fill pot
		fmt.Println("Replenishing pot")
		for i := 0; i < servingSize; i++ {
			pot <- 1
		}
		//Notify pot is full
		potFull <- 1
	}
}

func eat(potEmpty, pot, potFull chan int) {
	defer wg.Done()
	for {
		//Wait till the pot is full, ensures we don't eat untill cook fills it fully
		<-potFull
		select {
		case <-pot:
			//Eat from pot
			fmt.Printf("eating\n")
			potFull <- 1
			return
		default:
			//Pot is empty, need to notify cook
			potEmpty <- 1
		}
	}
}
